src/inference/trace_translators.jl

```diff
@@ -0,0 +1,874 @@
+        prev_model_trace, translator.p_new_args,
+        translator.argdiffs, constraints)
+
+    if !isempty(discard)
+        @error("can only extend the trace with random choices, cannot remove random choices")
+        error("Invalid SimpleExtendingTraceTranslator")
+    end
+
+    log_weight = log_model_weight - forward_proposal_score
+    return (new_model_trace, log_weight)
+end
+
+############################
+# SymmetricTraceTranslator #
+############################
+
+"""
+    translator = SymmetricTraceTranslator(;
+        q::GenerativeFunction,
+        q_args::Tuple = (),
+        f::TraceTransformDSLProgram)
+
+Constructor for a symmetric trace translator.
+
+Run the translator with:
+
+    (output_trace, log_weight) = translator(input_trace; check=false, observations=EmptyChoiceMap())
+
+Use `check` to enable the involution check (this requires that the transform `f` has been marked with [`is_involution`](@ref)).
+
+If `check` is enabled, then `observations` is a choice map containing the observed random choices, and the check will additionally ensure they are not mutated by the involution.
+"""
+@with_kw struct SymmetricTraceTranslator
+    q::GenerativeFunction
+    q_args::Tuple = ()
+    f::TraceTransformDSLProgram # an involution
+end
+
+function symmetric_trace_translator_run_transform(
+        f::TraceTransformDSLProgram,
+        prev_model_trace::Trace, forward_proposal_trace::Trace,
+        q::GenerativeFunction, q_args::Tuple)
+    first_pass_results = run_first_pass(f, prev_model_trace, forward_proposal_trace)
+    (new_model_trace, log_model_weight, _, discard) = update(
+        prev_model_trace, get_args(prev_model_trace),
+        map((_) -> NoChange(), get_args(prev_model_trace)),
+        first_pass_results.constraints)
+    log_abs_determinant = jacobian_correction(
+        f, prev_model_trace, forward_proposal_trace, first_pass_results, discard)
+    backward_proposal_trace, = generate(
+        q, (new_model_trace, q_args...), first_pass_results.u_back)
+    return (new_model_trace, backward_proposal_trace, log_abs_determinant)
+end
+
+function (translator::SymmetricTraceTranslator)(prev_model_trace::Trace; check=false, observations=EmptyChoiceMap())
+
+    # simulate from auxiliary program
+    forward_proposal_trace = simulate(translator.q, (prev_model_trace, translator.q_args...,))
+
+    # apply trace transform
+    (new_model_trace, backward_proposal_trace, log_abs_determinant) = symmetric_trace_translator_run_transform(
+        translator.f, prev_model_trace, forward_proposal_trace, translator.q, translator.q_args)
+
+    # compute log weight
+    prev_model_score = get_score(prev_model_trace)
+    new_model_score = get_score(new_model_trace)
+    forward_proposal_score = get_score(forward_proposal_trace)
+    backward_proposal_score = get_score(backward_proposal_trace)
+    log_weight = new_model_score - prev_model_score + backward_proposal_score + forward_proposal_score + log_abs_determinant
```

Comment:
At risk of asking a dumb question--if I'm doing my math correctly, wouldn't we want the `forward_proposal_score` to be negative?  Ie. `- forward_proposal_score` instead of `+ forward_proposal_score`?

Reply:
Since we want the acceptance ratio to be `(P[new_tr] * P[new_tr -> old_tr]) / (P[old_tr] * P[old_tr -> new_tr])`, where the `P[old_tr -> new_tr]` appears in the denominator.

Reply:
Yes, that's a bug. Thanks!